server/src/main/java/org/elasticsearch/index/codec/vectors/ES814ScalarQuantizedVectorsFormat.java

```diff
@@ -0,0 +1,123 @@
+
+import java.io.IOException;
+
+public class ES814ScalarQuantizedVectorsFormat extends FlatVectorsFormat {
+    public static final String QUANTIZED_VECTOR_COMPONENT = "QVEC";
+
+    static final String NAME = "ES814ScalarQuantizedVectorsFormat";
+
+    static final int VERSION_START = 0;
+    static final int VERSION_CURRENT = VERSION_START;
+    static final String META_CODEC_NAME = "Lucene99ScalarQuantizedVectorsFormatMeta";
+    static final String VECTOR_DATA_CODEC_NAME = "Lucene99ScalarQuantizedVectorsFormatData";
+    static final String META_EXTENSION = "vemq";
+    static final String VECTOR_DATA_EXTENSION = "veq";
+
+    private static final FlatVectorsFormat rawVectorFormat = new Lucene99FlatVectorsFormat();
+
+    /** The minimum confidence interval */
+    private static final float MINIMUM_CONFIDENCE_INTERVAL = 0.9f;
+
+    /** The maximum confidence interval */
+    private static final float MAXIMUM_CONFIDENCE_INTERVAL = 1f;
+
+    /**
+     * Controls the confidence interval used to scalar quantize the vectors the default value is
+     * calculated as `1-1/(vector_dimensions + 1)`
+     */
+    public final Float confidenceInterval;
+
+    public ES814ScalarQuantizedVectorsFormat(Float confidenceInterval) {
+        if (confidenceInterval != null
+            && (confidenceInterval < MINIMUM_CONFIDENCE_INTERVAL || confidenceInterval > MAXIMUM_CONFIDENCE_INTERVAL)) {
+            throw new IllegalArgumentException(
+                "confidenceInterval must be between "
+                    + MINIMUM_CONFIDENCE_INTERVAL
+                    + " and "
+                    + MAXIMUM_CONFIDENCE_INTERVAL
+                    + "; confidenceInterval="
+                    + confidenceInterval
+            );
+        }
+        this.confidenceInterval = confidenceInterval;
+    }
+
+    @Override
+    public String toString() {
+        return NAME + "(name=" + NAME + ", confidenceInterval=" + confidenceInterval + ", rawVectorFormat=" + rawVectorFormat + ")";
+    }
+
+    @Override
+    public FlatVectorsWriter fieldsWriter(SegmentWriteState state) throws IOException {
+        return new ES814ScalarQuantizedVectorsWriter(state, confidenceInterval, rawVectorFormat.fieldsWriter(state));
+    }
+
+    @Override
+    public FlatVectorsReader fieldsReader(SegmentReadState state) throws IOException {
+        return new ES814ScalarQuantizedVectorsReader(new Lucene99ScalarQuantizedVectorsReader(state, rawVectorFormat.fieldsReader(state)));
+    }
+
+    static class ES814ScalarQuantizedVectorsReader extends FlatVectorsReader {
+
+        final FlatVectorsReader reader;
+
+        ES814ScalarQuantizedVectorsReader(FlatVectorsReader reader) {
+            this.reader = reader;
+        }
+
+        @Override
+        public RandomVectorScorer getRandomVectorScorer(String field, float[] target) throws IOException {
+            return reader.getRandomVectorScorer(field, target);
+        }
```

Comment:
I understand not tackling it in this PR, but the random scorer returned here could use the native code stuff yes? The main cost here would be interacting with the on-heap `target`.

Reply:
Yes, that is certainly possible. Let's look into that in a follow up.